pywhy_graphs/algorithms/generic.py

```diff
@@ -333,3 +334,207 @@ def _single_shortest_path_early_stop(G, firstlevel, paths, cutoff, join, valid_p
+    """Recursively explores a graph to find a path.
+
+       Finds path that are compliant with the inducing path requirements.
+
+    Parameters
+    ----------
+    G : graph
+        The graph.
+    node_x : node
+        The source node.
+    node_y : node
+        The destination node
+    L : set
+        Set containing all the non-colliders.
+    S : set
+        Set containing all the colliders.
+    visited : set
+        Set containing all the nodes already visited.
+    xyancestors : set
+        Set containing the ancestors of X and Y.
+    cur_node : node
+        The current node.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path which is empty if the bool is false.
+    """
+    path_exists = False
+    path = []
+    visited.add(cur_node)
+    children = _find_directed_children(G, cur_node)
+
+    if cur_node is node_y:
+        return (True, [node_y])
+
+    for elem in children:
+        if elem in visited:
+            continue
+        else:
+            if (
+                _is_collider(G, elem)
+                and (elem not in S)
+                and (elem not in xyancestors)
+                and (elem is not node_y)
+            ):
+                continue
+            elif not _is_collider(G, elem) and (elem not in L) and (elem is not node_y):
+                continue
+            path_exists, temp_path = _recursive_path(
+                G, node_x, node_y, L, S, visited, xyancestors, elem
+            )
+            if path_exists:
+                path.append(cur_node)
+                path.extend(temp_path)
+                break
+    return (path_exists, path)
+
+
+def inducing_path(G, node_x, node_y, L=None, S=None):
+    """Checks if an inducing path exists between node_x and node_y and if it does returns it.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+    node_x : node
+        The source node.
+    node_y : node
+        The destination node.
+    L : set
+        The set containing every non-collider on the path. Defaults to an empty set.
+    S:  set
+        Contains all colliders on the path. Defaults to an empty set.
+
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path if the bool is true.
+    """
+
+    if L is None:
+        L = set()
+
+    if S is None:
+        S = set()
+
+    nodes = set(G.nodes)
+
+    if node_x not in nodes or node_y not in nodes:
+        raise ValueError("The provided nodes are not in the graph.")
+
+    if node_x == node_y:
+        raise ValueError("The start and destination nodes are the same.")
+
+    visited = set()
+    visited.add(node_x)
+
+    path = []  # this will contain the path.
+
+    xanc = _find_directed_parents(G, node_x)
+    yanc = _find_directed_parents(G, node_y)
+
+    xyancestors = xanc.union(yanc)
+
+    children = _find_directed_children(G, node_x)
```

Comment:
This is not correct:

```
    admg = ADMG()

    admg.add_edge("Y", "X", admg.directed_edge_name)
    admg.add_edge("Z", "Y", admg.bidirected_edge_name)
    admg.add_edge("Z", "H", admg.bidirected_edge_name)

    # this is an inducing path if L ={'Y'} and S = {}, yet it fails
    # X <- Y <-> z <-> H
    S = set()
    L = {'Y'}
    assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]
```

Reply:
Wait, in the graph that you constructed:


> # X <- Y <-> z <-> H

The directed edge is Y -> X . Is this a mistype or are you expecting a directed edge to mean a connection, no matter what way it is oriented? I thought an edge was two way traversable only if it was a bi-directed edge.

Reply:
Sorry this is a typo. I'll clarify on my computer later.

Reply:
```
# X <- Y <-> Z <-> H; Z -> X
    admg = ADMG()
    admg.add_edge("Y", "X", admg.directed_edge_name)
    admg.add_edge("Z", "X", admg.directed_edge_name)
    admg.add_edge("Z", "Y", admg.bidirected_edge_name)
    admg.add_edge("Z", "H", admg.bidirected_edge_name)

    # not an inducing path, since Y is not a collider and Y is not part of L
    S = {}
    L = {}
    assert not pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

    # now an inducing path, since Y is not a collider, but is part of L
    L = {'Y'}
    assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

    # X <-> Y <-> Z <-> H; Z -> X
    admg = ADMG()
    admg.add_edge("Y", "X", admg.bidirected_edge_name)
    admg.add_edge("Z", "X", admg.directed_edge_name)
    admg.add_edge("Z", "Y", admg.bidirected_edge_name)
    admg.add_edge("Z", "H", admg.bidirected_edge_name)

    # not an inducing path, since Y is not an ancestor of X, H, or S
    S = {}
    L = {}
    assert not pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

    # still not an inducing path, since Y is a collider
    L = {'Y'}
    assert not pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

    # now add an edge Y -> A
    admg.add_edge("Y", "A", admg.directed_edge_name)

    # an inducing path, since Y is a collider and is an ancestor of X, H, or S
    L = {}
    S = {'A'}
    assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

    # an inducing path, since Y is a collider and is an ancestor of X, H, or S
    L = {}
    S = {'Y'}
    assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]
```

These tests act as a lead up to various examples that should pass but does not based on the current implementation.

Reply:
>  now an inducing path, since Y is not a collider, but is part of L
>     L = {'Y'}
>     assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

@adam2392 There is still no directed path from X to H. Should this example still pass?

Reply:
>     # X <-> Y <-> Z <-> H; Z -> X
>     admg = ADMG()
>     admg.add_edge("Y", "X", admg.bidirected_edge_name)
>     admg.add_edge("Z", "X", admg.directed_edge_name)
>     admg.add_edge("Z", "Y", admg.bidirected_edge_name)
>     admg.add_edge("Z", "H", admg.bidirected_edge_name)
> 
>     # not an inducing path, since Y is not an ancestor of X, H, or S
>     S = {}
>     L = {}
>     assert not pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

Doesn't the bi-directed edge (X <-> Y) mean that Y is an ancestor of X? I have been taking it to mean that.

Reply:
> Doesn't the bi-directed edge (X <-> Y) mean that Y is an ancestor of X? I have been taking it to mean that.

No ancestor is only thru directed edges.

Reply:
> > now an inducing path, since Y is not a collider, but is part of L
> > L = {'Y'}
> > assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]
> 
> @adam2392 There is still no directed path from X to H. Should this example still pass?

According to the definition, yes:

```
# X <- Y <-> Z <-> H; Z -> X
```

Y is a non-collider and part of L. Z is a collider and an ancestor of X, Y, or S. X and H are endpoints. Lmk if that doesn't make sense considering the definition of inducing path.

Reply:
>     #  X <- Y <-> Z <-> H; Z -> X
>     admg = ADMG()
>     admg.add_edge("Y", "X", admg.directed_edge_name)
>     admg.add_edge("Z", "X", admg.directed_edge_name)
>     admg.add_edge("Z", "Y", admg.bidirected_edge_name)
>     admg.add_edge("Z", "H", admg.bidirected_edge_name)
> 
>     # not an inducing path, since Y is not a collider and Y is not part of L
>     S = {}
>     L = {}
>     assert not pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]
@adam2392 While I get the point about Y not being in L. But isn't there a path from X - Z - H? In this path, all the nodes obey the rules of inducing path. Unless, what you want is all the nodes on all the possible paths from X -> H have to obey the rules of an inducing path?

Reply:
Ah I see. In this case, X, Y, Z, H is not an inducing path, as discussed.

But if we look at X, Z, H, it is also not an inducing path. Z is not a collider and not part of L.

Reply:
I am a bit confused, earlier you said:
> A collider is a collider if there is two arrow endpoints. 

Z has two arrow endpoints pointing towards it. They are due to two bi-directed edges, so do I take it to mean that the two arrow endpoints have to be from directed edges?

Reply:
A collider is a node defined on a path. So it must have two arrow endpoints along the path. Here is something I found online that might help clarify: https://jeangoldinginstitute.blogs.bristol.ac.uk/2019/10/28/a-brief-introduction-to-colliders/

In (X, Z, H) path, Z is not a collider. In (X, Y, Z, H), Z is a collider on the path. Since colliders are defined wrt a path, that is the motivation for my comment on your function `is_collider(...)`, it must be defined wrt a triplet. 

E.g.

```
is_collider(X, Z, H) == False

is_collider(Y, Z, H) == True
```

Reply:
Aha! I finally understood what went wrong. Only the nodes on the path count! Looks like I will have to change the implementation slightly. But it can be done easily. Thanks for taking the time to explain this to me.

Reply:
One more question @adam2392. You said that any three nodes on the path can be checked if they form a collider, right?

> graph.isDefCollider(a, b, c)

This is what java does. a is the previous node, b is the current node and c is the iterator through all the neighbors of b (except a). It is only checking the collider status of a node w.r.t the node just before and just after it on the path.

Reply:
What is the question? If the collider is defined by 3 nodes? Yes that is a triplet.

Reply:
No I get that, what I am asking is, suppose that there is a path:

> A - B - C - D - F - E

When it comes to C, the java implementation is only checking if (B,C,D) form a collider. What if (B,C,F) do too? Will that not count as a collider?

Reply:
Yes but that's a different path? 

There could be a function that returns all possible inducing paths. And another that returns true/false and the first path it finds.

Reply:
> # an inducing path, since Y is a collider and is an ancestor of X, H, or S
>     L = {}
>     S = {'A'}
>     assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

@adam2392 Shouldn't this check be :
> assert not pywhy_graphs.inducing_path(admg, "X", "H",L,S)[0]

Or did you mean to add an edge between Y and some other node, not A?

Reply:
Which definition are you looking at? Can you link the paper, or copy/paste the text?

X is an endpoint,
Y is a collider and an ancestor of A, which is in S
Z is a collider and is an ancestor of X
H is an endpoint

therefore (X, Y, Z, H) is an inducing path relative to L and S.

Reply:
I read the definition wrong!

>and every collider on p is an ancestor of either X , Y , or a member of S

I thought this meant, every collider has to be an ancestor of X or Y, or a member of S.
But actually it has to be an ancestor of X, Y or a member of S.

Reply:
@adam2392 by the same definition, in this example:


>     an inducing path, since Y is a collider and is an ancestor of X, H, or S
>     L = {}
>     S = {'Y'}
>     assert pywhy_graphs.inducing_path(admg, "X", "H", L, S)[0]

Isn't why a collider but not an ancestor of any element in S or X or Y? (It is itself in S)

Reply:
Read the preliminaries section in Zhang 2008 (https://www.sciencedirect.com/science/article/pii/S0004370208001008). By convention A is an ancestor of A.

Reply:
Ah. That does make sense. I kind of assumed this was the case and made the modification according to it in the last commit. All the test cases you provided pass. Do you want to take a look at the implementation again?